src/plugins/data_source_management/public/components/data_source_selector/data_source_selector.tsx

```diff
@@ -57,66 +58,130 @@ export class DataSourceSelector extends React.Component<
     this.state = {
       allDataSources: [],
       defaultDataSource: '',
-      selectedOption: this.props.defaultOption
-        ? this.props.defaultOption
-        : this.props.hideLocalCluster
-        ? []
-        : [LocalCluster],
+      selectedOption: this.props.hideLocalCluster ? [] : [LocalCluster],
     };
   }
 
   componentWillUnmount() {
     this._isMounted = false;
   }
 
-  async componentDidMount() {
-    this._isMounted = true;
+  handleSelectedOption(
+    dataSourceOptions: DataSourceOption[],
+    allDataSources: Array<SavedObject<DataSourceAttributes>>,
+    defaultDataSource: string | null
+  ) {
+    const [{ id }] = this.props.defaultOption!;
+    const dataSource = dataSourceOptions.find((ds) => ds.id === id);
+    const selectedOption = dataSource ? [{ id, label: dataSource.label }] : [];
+
+    // Invalid/filtered out datasource
+    if (!dataSource) {
+      this.props.notifications.addWarning(
+        i18n.translate('dataSource.fetchDataSourceError', {
+          defaultMessage: 'Data source with id is not available',
+        })
+      );
+    }
 
-    const currentDefaultDataSource = this.props.uiSettings?.get('defaultDataSource', null) ?? null;
     this.setState({
       ...this.state,
-      defaultDataSource: currentDefaultDataSource,
+      dataSourceOptions,
+      selectedOption,
+      defaultDataSource,
+      allDataSources,
     });
+    this.props.onSelectedDataSource(selectedOption);
+  }
```

Comment:
What if plugins team pass in local cluster?

Reply:
`Local Cluster` should already be in the `dataSources` because we check if the `hideLocalCluster` is disabled and add the local cluster before calling this method. In this case, the `id = ''` and because `Local Cluster` is in the list of datasources, `dataSources.find((ds) => ds.id === id)` should return the `Local Cluster`

Reply:
Make sense!